File: java/socketd/src/main/java/org/noear/socketd/transport/stream/impl/StreamBase.java
package org.noear.socketd.transport.stream.impl;

import org.noear.socketd.exception.SocketDTimeoutException;
import org.noear.socketd.transport.stream.Stream;
import org.noear.socketd.transport.stream.StreamInternal;
import org.noear.socketd.transport.stream.StreamManger;
import org.noear.socketd.utils.RunUtils;
import org.noear.socketd.utils.TriConsumer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.ScheduledFuture;
import java.util.function.Consumer;

/**
 * 流基类
 *
 * @author noear
 * @since 2.0
 */
public abstract class StreamBase<T extends Stream> implements StreamInternal<T> {
    private static final Logger log = LoggerFactory.getLogger(Stream.class);

    //保险任务
    private ScheduledFuture<?> insuranceFuture;

    private final String sid;
    private final int demands;

    private long timeout;
    protected Consumer<Throwable> doOnError;
    protected TriConsumer<Boolean, Integer, Integer> doOnProgress;

    public StreamBase(String sid, int demands, long timeout) {
        this.sid = sid;
        this.demands = demands;
        this.timeout = timeout;
    }

    @Override
    public String sid() {
        return sid;
    }

    /**
     * 需求数量（0，1，2）
     */
    @Override
    public int demands() {
        return demands;
    }

    /**
     * 配置超时
     */
    public T timeout(long timeout){
        this.timeout = timeout;
        return (T)this;
    }

    /**
     * 超时
     */
    @Override
    public long timeout() {
        return timeout;
    }

    /**
     * 保险开始（避免永久没有回调，造成内存不能释放）
     *
     * @param streamManger  流管理器
     * @param streamTimeout 流超时
     */
    @Override
    public void insuranceStart(StreamManger streamManger, long streamTimeout) {
        if (insuranceFuture != null) {
            return;
        }

        insuranceFuture = RunUtils.delay(() -> {
            streamManger.removeStream(sid);
            this.onError(new SocketDTimeoutException("The stream response timeout, sid=" + sid));
        }, streamTimeout);
    }

    /**
     * 保险取消息
     */
    @Override
    public void insuranceCancel() {
        if (insuranceFuture != null) {
            insuranceFuture.cancel(false);
        }
    }

    @Override
    public void onError(Throwable error) {
        if (doOnError != null) {
            doOnError.accept(error);
        } else {
            if (log.isDebugEnabled()) {
                log.debug("The stream error, sid={}", sid(), error);
            }
        }
    }

    @Override
    public void onProgress(boolean isSend, int val, int max) {
        if (doOnProgress != null) {
            doOnProgress.accept(isSend, val, max);
        }
    }

    @Override
    public T thenError(Consumer<Throwable> onError) {
        this.doOnError = onError;
        return (T) this;
    }

    @Override
    public T thenProgress(TriConsumer<Boolean, Integer, Integer> onProgress) {
        this.doOnProgress = onProgress;
        return (T) this;
    }
}

File: python/socketd/socketd/transport/core/stream/StreamBase.py
import asyncio
from abc import ABC
from typing import Callable, Optional

from socketd.exception.SocketdTimeoutException import SocketdTimeoutException
from socketd.transport.core.stream.StreamManger import StreamInternal, StreamManger
from socketd.transport.utils.CompletableFuture import CompletableFuture


class StreamBase(StreamInternal, ABC):
    """流接收器基类"""

    def __init__(self, sid: str, demands: int, timeout: int):
        self.__sid = sid
        self.__timeout = timeout
        self.__demands = demands
        self.__doOnError: Callable[[Exception], None] = None
        self.__doOnProgress: Callable[[bool, int, int], None] = None
        self.__insuranceFuture: Optional[asyncio.Future] = None

    def on_error(self, error: Exception):
        if error:
            self.__doOnError(error)

    def is_done(self):
        return True

    def then_progress(self, on_progress: Callable[[bool, int, int], None]):
        pass

    def get_sid(self) -> str:
        return self.__sid

    def timeout(self):
        return self.__timeout

    def then_error(self, onError: Callable[[Exception], None]):
        self.__onError = onError

    def demands(self) -> int:
        return self.__demands

    def insurance_start(self, streamManger: StreamManger, streamTimeout: float) -> None:
        if self.__insuranceFuture:
            return

        async def __insuranceFuture():
            await asyncio.sleep(streamTimeout)
            streamManger.remove_stream(self.__sid)
            self.on_error(SocketdTimeoutException(f"The stream response timeout, sid={self.__sid}"))

        self.__insuranceFuture = CompletableFuture(__insuranceFuture())
        asyncio.run_coroutine_threadsafe(self.__insuranceFuture.get(streamTimeout), asyncio.get_running_loop())

    def insurance_cancel(self) -> None:
        if self.__insuranceFuture:
            self.__insuranceFuture.cancel()

    def on_progress(self, is_send, val, max_val):
        if self.__doOnProgress:
            self.__doOnProgress(is_send, val, max_val)
